### PR TITLE
Add support for validate_vms action on transformation_mappings

### PR DESCRIPTION
**Pull request by abellotti**
Extends https://github.com/ManageIQ/manageiq-api/pull/352

Dependent on https://github.com/ManageIQ/manageiq/pull/17177

Supports import from csv or without any data at all

```
POST /api/transformation_mappings/:id
{
   "action": "validate_vms"
}
```

```
POST /api/transformation_mappings/:id
{
   "action": "validate_vms",
   "import": [
      { "name": "vm_name", "uid": "vm_uid" } 
    ]
}
```

cc: @gtanzillo @bzwei

### CODE REVIEW

**Comment by abellotti** [comment]
Marking as WIP as we need https://github.com/ManageIQ/manageiq/pull/17177 merged first

resolves issue https://github.com/ManageIQ/manageiq-api/pull/352#issuecomment-378406745

so results now would return vms hrefs, so:

```
{
    "valid_vms": [
        {
            "href": "http://localhost:3000/api/vms/2222",
            "name": "amaya-at",
            "cluster": "Default",
            "path": "RHV/Default",
            "allocated_size": "32212254720.0",
            "id": "2222",
            "reason": "OK"
        }
    ]
}
```

**Comment by abellotti** [comment]
@gmcculloug had to update the specs to have the proper wiring for the valid vm case.

ping @imtayadeway can we borrow your  👀 for review ? Thanks.

File: app/controllers/api/transformation_mappings_controller.rb [hunk 1]
```diff
@@ -10,6 +10,13 @@ def create_resource(_type, _id, data = {})
       raise BadRequestError, "Could not create Transformation Mapping - #{err}"
     end
 
+    def validate_vms_resource(type, id, data = {})
+      transformation_mapping = resource_search(id, type, collection_class(type))
+      normalize_hash(:vms, transformation_mapping.validate_vms(data["import"]) || {})
+    rescue StandardError => err
+      raise BadRequestError, "Could not validate vms - #{err}"
```
[the remark this answers is not in the document]
Comment (abellotti) [reply app/controllers/api/transformation_mappings_controller.rb]: Not sure we can anticipate what can wrong, this is outside API's knowledge and the validate_vms does not raise any malformed input error for now, so really nothing to catch and make a BadRequestErrorr of.  we should probably just drop this rescue and resort to the 500.

File: app/controllers/api/transformation_mappings_controller.rb [hunk 2]
```diff
@@ -10,6 +10,13 @@ def create_resource(_type, _id, data = {})
       raise BadRequestError, "Could not create Transformation Mapping - #{err}"
     end
 
+    def validate_vms_resource(type, id, data = {})
+      transformation_mapping = resource_search(id, type, collection_class(type))
+      normalize_hash(:vms, transformation_mapping.validate_vms(data["import"]) || {})
```
[the remark this answers is not in the document]
Comment (abellotti) [reply app/controllers/api/transformation_mappings_controller.rb]: This is fine too for now, but there is still an issue with the normalize_ methods where type is not properly propagated and type is deduced from the request when we explicitly are telling normalize_hash the type.

**Comment by abellotti** [comment]
updated & re-pushed @imtayadeway

/cc @AparnaKarve

**Comment by gtanzillo** [comment]
@imtayadeway Can we get this one merged? If there are additional changes that we need, we can do that in a followup PR.

File: app/controllers/api/transformation_mappings_controller.rb [hunk 3]
```diff
@@ -10,6 +10,18 @@ def create_resource(_type, _id, data = {})
       raise BadRequestError, "Could not create Transformation Mapping - #{err}"
     end
 
+    def validate_vms_resource(type, id, data = {})
+      transformation_mapping = resource_search(id, type, collection_class(type))
+      (transformation_mapping.validate_vms(data["import"]) || {}).tap do |res|
+        %w(valid_vms invalid_vms conflict_vms).each do |key|
+          next unless res.key?(key)
+          res[key].each do |entry|
+            entry["href"] = normalize_href(:vms, entry["id"]) if entry["href"].blank? && entry["id"].present?
```
Comment (imtayadeway) [inline app/controllers/api/transformation_mappings_controller.rb]: I'm not sure why this postfix conditional is needed - isn't it part of the contract of the return value of `validate_vms` that it *will* have an id, and that it *won't* have an href? If there are cases where this is needed, this should have a test. Perhaps you could handle this in a follow up PR? Thanks @abellotti !

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit eeeb1992d5ff128ef170517d20d4ee7a50f0ff07
Author: Tim Wade <imtayadeway@users.noreply.github.com>
Date:   Mon Apr 30 07:40:47 2018 -0700

    Merge pull request #358 from abellotti/transformation_mapping_validate_vms
    
    Add support for validate_vms action on transformation_mappings
    (cherry picked from commit 94370162a7b6e7e6d896fac21f7cc2b09747c127)
```

**Pull request merged**